clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5910,6 +5910,30 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
   D->addAttr(::new (S.Context) BuiltinAliasAttr(S.Context, AL, Ident));
 }
 
+static void handleDebugInfoTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  if (!AL.hasParsedType()) {
+    S.Diag(AL.getLoc(), diag::err_attribute_wrong_number_arguments) << AL << 1;
+    return;
+  }
+
+  TypeSourceInfo *ParmTSI = nullptr;
+  QualType type = S.GetTypeFromParser(AL.getTypeArg(), &ParmTSI);
+  assert(ParmTSI && "no type source info for attribute argument");
+
+  if (type->isEnumeralType()) {
+    QualType BitfieldType = llvm::cast<FieldDecl>(D)->getType();
+    QualType EnumUnderlyingType =
+        type->getAs<EnumType>()->getDecl()->getIntegerType();
+    if (EnumUnderlyingType != BitfieldType) {
```

Comment:
> A few things -- I think we should look at the canonical type so that we don't run into issues with typedefs. e.g.,
> 
> ```
> enum E { Zero, One };
> typedef int Foo;
> struct S {
>   [[clang::debug_info_type(E) Foo field : 1;
> };
> ```
> 
> where the enum's underlying type is `int` but the bit-field's type is a typedef.
> 
> Another case we should consider would be whether we want to allow signed unsigned mismatches, as in:
> 
> ```
> enum E { Invalid = -1, ValidValue, OtherValidValue, YetAnotherValidValue };
> struct S {
>   [[clang::debug_info_type(E) unsigned field : 3;
> };
> ```
> 
> and only prevent situations where the specified type is not an enumeration or integral type.

This ALSO makes me wonder if forcing it to be an enum type is necessary.  I can see potential value of a "OpaqueValueOfStruct" storage type thing, that perhaps we should just 'trust' that the user is doing something sensible with it.

  I think in the signed/unsigned mismatch, we should allow it (thanks to the layout rules).  That is, consider:

```
enum SIGNED {Invalid = -1, Valid };
enum UNSIGNED {Valid, INVALID = 0xF };

struct S {
  [[attr_name(SIGNED)]]
  unsigned SignedThing : 4;
  [[attr_name(UNSIGNED)]]
  unsigned UnsignedThing : 4;
};
```
They HAVE to be the same type else they don't get packed into the same byte, but you have 1 enum that can represent a negative value.

Reply:
> and only prevent situations where the specified type is not an enumeration or integral type.

> This ALSO makes me wonder if forcing it to be an enum type is necessary.

I intentionally haven't been restricting the set of types that can be passed as the argument to `preferred_type`. This makes even more sense now that we are going with more generic name. I'm just making sure that enum-related diagnostics are triggered only when enum was passed by the user.

> Another case we should consider would be whether we want to allow signed unsigned mismatches:

> I think in the signed/unsigned mismatch, we should allow it (thanks to the layout rules)

> 1 enum that can represent a negative value

I disagree. We shouldn't pretend that our enums can hold negative values, while we store them in unsigned bit-fields (save for rare exceptions, if any). Now that we are not bound by C++98 compilers, I believe we should propagate underlying type of our enums from bit-field type they are stored in (usually `unsigned`, `uint64_t` for `DeclBase.h`), and deal with negative values at enumerator declaration (I guess we'll have to explicitly cast them to the underlying type). I can do that while adding `clang::preferred_type` to bit-fields.

Ideally I'd propagate bit-field width as well, but that doesn't seem possible at the moment: https://github.com/llvm/llvm-project/pull/69104#issuecomment-1770189229

Reply:
I misread an earlier comment of yours (I thought you were saying that the underlying type of the enum and the bit-field needed to match, not that we're pulling information from one and applying it to the other.)

Reply:
I think we should ignore signed vs unsigned mismatch on the enum vs bit-field types because I'm concerned about how chatty the diagnostic will be with portable code. MSVC and Clang/GCC make different decisions about whether an unscoped enumeration has a signed or unsigned underlying type, so it makes me wary.

We can always revisit this decision in the future if we find a better heuristic or more worrisome code in the wild.

WDYT?

Reply:
After chatting with Aaron offline, I used his solution to ignore sign-ness when comparing bit-field type against enum enumeration type.

Reply:
With `typedef` and sign-ness issues fixed, I consider this done.